### ISSUE TITLE
Returned type of find() should include only the selected fields

### ISSUE THREAD

**Issue by fingeromer** [typeorm/typeorm#10034]
### Feature Description

The `find()` method allows to specify which fields to query by using `select` option:
```
const items = repository.find({
      relations: { ... },
      select: {
        a: true,
        b: true,
        cArray: false
      },
      where: { ... },
    });
```

While each object in the returned `items` array only carries the selected 'a' and 'b' properties at runtime, the TypeScript compiler doesn't have any knowledge of this. It will allow you to access any other properties defined on the entity after the query.
It might be dangerous, because accessing 'cArray' in runtime might cause something like:
`TypeError: Cannot read property 'length' of undefined`

Fixing this will make TypeORM much more type safe, without this mismatch between the actual returned value, to the return type of the function.

### The Solution

Looking the example above, I would expect the returned type to be `Pick<EntityA, 'a', 'b'>`

### Considered Alternatives

If it's too hard to do this in the existing `find()` (and also causes breaking changes) another possibility is to add a new `select()` function.

### Additional Context

https://www.prisma.io/docs/concepts/more/comparisons/prisma-and-typeorm#selecting-fields

### Relevant Database Driver(s)

- [ ] aurora-mysql
- [ ] aurora-postgres
- [ ] better-sqlite3
- [ ] cockroachdb
- [ ] cordova
- [ ] expo
- [ ] mongodb
- [ ] mysql
- [ ] nativescript
- [ ] oracle
- [X] postgres
- [ ] react-native
- [ ] sap
- [ ] spanner
- [ ] sqlite
- [ ] sqlite-abstract
- [ ] sqljs
- [ ] sqlserver

### Are you willing to resolve this issue by submitting a Pull Request?

Yes, I have the time, and I know how to start.

**Comment by iJhefe** [comment]
I agree with this but it may can have change severals breaking changes in who is using full entity return type and is selecting partial

**Comment by aberonni** [comment]
I've been following this issue as I believe it to be one of the bigger pain points in typing in typeorm as it stands. I would like to see it reopened and prioritised.

**Comment by jonaskordt** [comment]
I would like to see the same thing (deeply) for relations which are being loaded. However, I'm not sure how this would work with eagerly loaded relations since they are present even if they are not explicitly selected.

**Comment by Nosfistis** [comment]
In strict typing I use this convention:

```Typescript
class MyEntity {

@OneToMany(() => RelationA, r => r.my, { eager: true })
relationA!: RelationA;

@OneToMany(() => RelationB, r => r.my)
relationB?: RelationB;
```

This way typing assumes that relation A is always there.

**Comment by jonaskordt** [comment]
> In strict typing I use this convention:
> 
> class MyEntity {
> 
> @OneToMany(() => RelationA, r => r.my, { eager: true })
> relationA!: RelationA;
> 
> @OneToMany(() => RelationB, r => r.my)
> relationB?: RelationB;
> This way typing assumes that relation A is always there.

I know this works. But it means if I do load a non-eager relation I still need to either to a null check before accessing it or to a non-null assertion. It would be great if the return value of `find` would already do this based on the relations which were loaded.

**Comment by Nosfistis** [comment]
Exactly, the proposition of this issue is when loading a non-eager relation using `FindOptions` then the resulting type adds `Required` to those relations, removing the requirement for non-null assertions.
